### PR TITLE
core(user-flow): start/end navigation functions

### PR DESCRIPTION
**Pull request by adamraine**
We previously discussed this idea 🔒 [here](http://go/user-triggered-navigations). Using some promise/callback BS we can implement both the callback approach and start/end approach for user triggered navigations.

### CODE REVIEW

File: lighthouse-core/test/fraggle-rock/scenarios/start-end-navigation-test-pptr.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    state.server.baseDir = `${LH_ROOT}/lighthouse-core/test/fixtures/fraggle-rock/navigation-basic`;
+  });
+
+  it('should capture a navigation via user interaction', async () => {
+    const pageUrl = `${state.serverBaseUrl}/links-to-index.html`;
+    await state.page.goto(pageUrl, {waitUntil: ['networkidle0']});
+
+    const flow = await lighthouse.startFlow(state.page);
+
+    await flow.startNavigation();
```
Comment (connorjclark) [inline lighthouse-core/test/fraggle-rock/scenarios/start-end-navigation-test-pptr.js:33]: Do we need docs?

**Review by connorjclark — APPROVED** [review]
+1, but docs?

**Comment by adamraine** [comment]
Yeah docs sound good. @paulirish is https://github.com/GoogleChrome/lighthouse/pull/14021 ready to go first?

Also, just had an idea. We could add a new smoke test runner for this. I'll do it in a separate PR.

File: lighthouse-core/fraggle-rock/user-flow.js [hunk 2]
```diff
@@ -108,20 +109,60 @@ class UserFlow {
...
 
     this._addGatherStep(gatherResult, options);
+  }
+
+  /**
+   * @param {StepOptions=} stepOptions
+   */
+  async startNavigation(stepOptions) {
+    /** @type {ExternalPromise<() => void>} */
+    const setupPromise = new ExternalPromise();
```
Comment (paulirish) [inline lighthouse-core/fraggle-rock/user-flow.js:125]: why can't this just be a promise?  looking at the extpromise impl i'm not seeing what's special about it…

File: lighthouse-core/fraggle-rock/user-flow.js [hunk 3]
```diff
@@ -108,20 +109,60 @@ class UserFlow {
...
+      if (this.currentNavigation) {
+        // If the navigation already started, re-throw the error so it is emitted when `navigatePromise` is awaited.
+        throw err;
+      } else {
+        // If the navigation has not started, reject the `setupPromise` so the error throws when it is awaited in `startNavigation`.
+        setupPromise.reject(err);
+      }
+    });
+
+    const continueNavigation = await setupPromise;
```
Comment (paulirish) [inline lighthouse-core/fraggle-rock/user-flow.js:143]: can you rename `setupPromise` to something more descriptive?

File: lighthouse-core/fraggle-rock/user-flow.js [hunk 4]
```diff
@@ -108,20 +109,60 @@ class UserFlow {
...
     const options = this._getNextNavigationOptions(stepOptions);
     const gatherResult = await navigationGather(requestor, options);
 
     this._addGatherStep(gatherResult, options);
+  }
+
+  /**
+   * @param {StepOptions=} stepOptions
+   */
+  async startNavigation(stepOptions) {
```
Comment (paulirish) [inline lighthouse-core/fraggle-rock/user-flow.js:123]: add a comment describing that this is an alternative to the `.navigate()` method? and perhaps linking to the docs you  added?

File: lighthouse-core/fraggle-rock/user-flow.js (hunk 2 above)
Comment (connorjclark) [reply lighthouse-core/fraggle-rock/user-flow.js:125]: There's a big amount of indirection here that takes a moment to figure out. I tried to help in my comment suggestion.

Comment (adamraine) [reply lighthouse-core/fraggle-rock/user-flow.js:125]: The main goal was to make the code easier to read by reducing the nested promises and such. It was a generic pattern that we might user later, but that's just wishful thinking I guess.

I'm fine going with paul's suggestion here.

**Review by paulirish — APPROVED** [review]

**Pull request merged**